parts/xmem_nozzle.go

```diff
@@ -1994,11 +1913,26 @@ func (xmem *XmemNozzle) uncompressBody(req *base.WrappedMCRequest) error {
 	return nil
 }
 func (xmem *XmemNozzle) updateSystemXattrForTarget(wrappedReq *base.WrappedMCRequest, lookup *base.SubdocLookupResponse, setOpt SetMetaXattrOptions) (err error) {
-	if !setOpt.sendHlv && !setOpt.preserveSync {
+	var needToUpdateSysXattrs bool
+	if wrappedReq.Req.DataType&mcc.XattrDataType != 0 {
+		// in this case, we may need to update HLV if source already has it, even if mobile/HLV are all off.
+		needToUpdateSysXattrs = true
+	} else if setOpt.preserveSync {
+		needToUpdateSysXattrs = true
+	} else if xmem.source_cr_mode == base.CRMode_Custom {
+		needToUpdateSysXattrs = true
+	} else if setOpt.sendHlv {
+		maxCas := xmem.config.vbMaxCas[wrappedReq.Req.VBucket]
+		if wrappedReq.Req.Cas >= maxCas {
+			needToUpdateSysXattrs = true
+		}
+	}
```

Comment:
And maybe one more case of if the target document already has HLV? Even if maxCas condition was not met?

Reply:
This is covered in line 1917. If there is XATTR, needToUpdateSysXattrs is true. Later in the same routine we call call crMeta.NeedToUpdateHlv() to check for HLV and do the update if needed.